common/clock/ratelimiter.go

```diff
@@ -0,0 +1,418 @@
+		//    This is equivalent to calling Cancel() on a [rate.Reservation], but it is more likely
+		//    to return the token.
+		Used(wasUsed bool)
+	}
+
+	ratelimiter struct {
+		// important concurrency note: the lock MUST be held while acquiring AND handling Now(),
+		// to ensure that times are handled monotonically.
+		// otherwise, even though `time.Now()` is monotonic, they may make progress
+		// past the mutex in a different order and no longer be handled in a monotonic way.
+		timesource TimeSource
+		latestNow  time.Time // updated on each call, never allowed to rewind
+		limiter    *rate.Limiter
+		mut        sync.Mutex
+	}
+
+	// a reservation that allows an immediate call, and can be canceled
+	allowedReservation struct {
+		res *rate.Reservation
+		// reservedAt is used to cancel at the reservation time, restoring tokens
+		// if (and only if) no other time-advancing calls have interleaved
+		reservedAt time.Time
+		// needs access to the parent-wrapped-ratelimiter to ensure time is not rewound
+		limiter *ratelimiter
+	}
+	// a reservation that does NOT allow an immediate call.
+	// the *rate.Reservation has already been canceled, so this type is trivial.
+	deniedReservation struct{}
+)
+
+var (
+	_ Ratelimiter = (*ratelimiter)(nil)
+	_ Reservation = (*allowedReservation)(nil)
+	_ Reservation = deniedReservation{}
+)
+
+// err constants to keep allocations low
+var (
+	// ErrCannotWait is used as the common base for two internal reasons why Ratelimiter.Wait call cannot
+	// succeed, without directly exposing the reason except as an explanatory string.
+	//
+	// Callers should only ever care about ErrCannotWait to be resistant to racing behavior.
+	// Or perhaps ideally: not care about the reason at all beyond "not a ctx.Err()".
+	ErrCannotWait             = fmt.Errorf("ratelimiter.Wait() cannot be satisfied")
+	errWaitLongerThanDeadline = fmt.Errorf("would wait longer than ctx deadline: %w", ErrCannotWait)
+	errWaitZeroBurst          = fmt.Errorf("zero burst will never allow: %w", ErrCannotWait)
+)
+
+func NewRatelimiter(lim rate.Limit, burst int) Ratelimiter {
+	return &ratelimiter{
+		timesource: NewRealTimeSource(),
+		limiter:    rate.NewLimiter(lim, burst),
+	}
+}
+
+func NewMockRatelimiter(ts TimeSource, lim rate.Limit, burst int) Ratelimiter {
+	return &ratelimiter{
+		timesource: ts,
+		limiter:    rate.NewLimiter(lim, burst),
+	}
+}
+
+// lockNow updates the current "now" time, and ensures it never rolls back.
+// This should be equivalent to setting `r.timesource.Now()`, outside of tests
+// that abuse mocked time and Advance with negative values.
+//
+// The lock MUST be held until all time-accepting methods are called on the
+// underlying rate.Limiter, as otherwise the internal "now" value may rewind
+// and cause undefined behavior.
+func (r *ratelimiter) lockNow() (now time.Time, unlock func()) {
+	r.mut.Lock()
+	// important design note: `r.timesource.Now()` MUST be gathered while the lock is held,
```

Comment:
good point, but repeated above

Reply:
I had intended this as "external use needs to hold the lock as long as the limiter is being modified (often handled by `defer cancel()`)" being a different concern from "this func's implementation must lock before gathering `time.Now()`" (which can't be observed by callers of `lockNow()`).  "do less inside the lock" -> "move Now() outside the lock" is a fairly predictable future optimization thought, but it'd reproduce the underlying `*rate.Limiter` flaw.

I can reword it, but I do think both need to be documented explicitly, in a way that'd be noticed during code review for any changes.

Reply:
ah, wait, you were probably referring to the comment above `timesource` in the struct definition.  yea, that feels like a duplicate.  I'll clean that up.